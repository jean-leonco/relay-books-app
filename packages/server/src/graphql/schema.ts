import { GraphQLSchema } from 'graphql';

import applyMiddlewares from '../core/middlewares/applyMiddlewares';

// import SubscriptionType from './subscription/SubscriptionType';
import MutationType from './type/MutationType';
import QueryType from './type/QueryType';

const _schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
  // subscription: SubscriptionType,
});

applyMiddlewares(_schema);

export const schema = _schema;
