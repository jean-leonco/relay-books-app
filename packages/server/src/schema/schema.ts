import { GraphQLSchema } from 'graphql';

import applyMiddlewares from '../graphql/middlewares/applyMiddlewares';

import MutationType from './MutationType';
import QueryType from './QueryType';

const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});

applyMiddlewares(schema);

export default schema;
