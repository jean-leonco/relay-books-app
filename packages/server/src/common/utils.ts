import { Model, Types } from 'mongoose';
import { fromGlobalId } from 'graphql-relay';

// https://stackoverflow.com/a/2593661/710693
export const escapeRegex = (str: string) => `${str}`.replace(/[.?*+^$[\]\\(){}|-]/g, '\\$&');

export const PLATFORM = {
  APP: 'APP',
  UNKNOWN: 'UNKNOWN',
};

export const getPlatform = (platform?: string | null): string => {
  if (typeof platform !== 'string') {
    return PLATFORM.UNKNOWN;
  }

  //if (platform.includes('ios') || platform.includes('android')) {
  //  return PLATFORM.APP;
  //}

  if (platform === PLATFORM.APP) {
    return PLATFORM.APP;
  }

  return PLATFORM.UNKNOWN;
};

export function kilometersToEarthRadians(kilometers: number) {
  return kilometers / 6378.1;
}

// returns an ObjectId given an param of unknown type
export const getObjectId = (target: string | Model<any> | Types.ObjectId | undefined | null): Types.ObjectId | null => {
  if (!target) {
    return null;
  }

  if (target instanceof Types.ObjectId) {
    return new Types.ObjectId(target.toString());
  }

  if (typeof target === 'object') {
    return target && target._id ? new Types.ObjectId(target._id) : null;
  }

  if (typeof target === 'string') {
    const result = fromGlobalId(target);

    if (result.type && result.id && Types.ObjectId.isValid(result.id)) {
      return new Types.ObjectId(result.id);
    }

    if (Types.ObjectId.isValid(target)) {
      return new Types.ObjectId(target);
    }

    return null;
  }

  return null;
};

type EnumObject = {
  [key: string]: string;
};

type EnumObjectResult = {
  [key: string]: {
    value: string;
  };
};

export const enumBuilderValues = (constants: EnumObject): EnumObjectResult =>
  Object.keys(constants).reduce(
    (prev, curr) => ({
      ...prev,
      [curr]: {
        value: constants[curr],
      },
    }),
    {},
  );
