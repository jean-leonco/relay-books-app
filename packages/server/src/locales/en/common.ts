const common = {
  SomethingWentWrong: 'Something went wrong.',
  TheDescriptionShouldNot: 'The description must be less than {{size}} characters.',
};
export default common;
export type Keys = keyof typeof common;
