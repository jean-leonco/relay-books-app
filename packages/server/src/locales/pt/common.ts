const common = {
  SomethingWentWrong: 'Algo deu errado.',
  TheDescriptionShouldNot: 'A descrição deve ter menos de {{size}} caracteres.',
};
export default common;
export type Keys = keyof typeof common;
