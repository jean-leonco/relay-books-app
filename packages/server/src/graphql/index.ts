/* eslint-disable no-console */
import 'core-js/stable';
import { createServer } from 'http';

import { execute, subscribe } from 'graphql';
import { SubscriptionServer } from 'subscriptions-transport-ws';

import { GRAPHQL_PORT, GRAPHQL_HOST } from '../common/config';
import connectDatabase from '../common/database';

import app from './app';
import { schema } from './schema';

type ConnectionParams = {
  authorization?: string;
};

const runServer = async () => {
  try {
    console.log('\n🔗 Connecting to database...');
    await connectDatabase();
  } catch (error) {
    console.error('Could not connect to database', { error });
    throw error;
  }

  const server = createServer(app.callback());

  server.listen(GRAPHQL_PORT, () => {
    console.info(`\n🚀 Server started at http://${GRAPHQL_HOST}:${GRAPHQL_PORT}`);

    if (process.env.NODE_ENV !== 'production') {
      console.info(`\n🎠 GraphQL Playground available at http://${GRAPHQL_HOST}:${GRAPHQL_PORT}/playground\n`);
    }

    SubscriptionServer.create(
      {
        onConnect: async (connectionParams: ConnectionParams) => {
          console.info('Client subscription connected', connectionParams);
          // @TODO - get user from connectionParams?.authorization
          // @TODO - return context with user
        },
        onDisconnect: () => console.info('Client subscription disconnected'),
        execute,
        subscribe,
        schema,
      },
      {
        server,
        path: '/subscriptions',
      },
    );
  });
};

(async () => {
  console.log('\n📡 Server starting...');
  await runServer();
})();
