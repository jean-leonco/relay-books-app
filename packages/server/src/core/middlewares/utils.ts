import { GraphQLResolveInfo } from 'graphql';
import { ObjectSchema, Shape } from 'yup';

import { GraphQLContext } from '../../types';

interface GraphQLMutationConfig {
  validationSchema?(context: GraphQLContext): ObjectSchema<Shape<any, any>>;
  authenticatedOnly?: boolean;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): GraphQLMutationConfig | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
