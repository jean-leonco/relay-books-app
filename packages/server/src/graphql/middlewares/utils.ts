import { GraphQLField, GraphQLResolveInfo } from 'graphql';

import { Extensions } from '../../types';

interface Field extends GraphQLField<any, any> {
  extensions: Extensions | null | undefined;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): Field | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
