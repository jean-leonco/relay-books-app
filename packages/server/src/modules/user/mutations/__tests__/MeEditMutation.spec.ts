import { graphql } from 'graphql';

import { sanitizeTestObject } from '@booksapp/test-utils';

import { schema } from '../../../../graphql/schema';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  createUser,
  disconnectMongoose,
  getContext,
  gql,
} from '../../../../../test/helpers';
import { PLATFORM } from '../../../../common/utils';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('MeEditMutation', () => {
  it('should edit me', async () => {
    const user = await createUser();

    const name = 'Awesome Name';
    const email = 'email@newmail.com';

    const mutation = gql`
      mutation M($input: MeEditInput!) {
        MeEdit(input: $input) {
          me {
            id
            fullName
            email
          }
          error
        }
      }
    `;

    const variables = {
      input: {
        name,
        email,
      },
    };
    const rootValue = {};
    const context = await getContext({ user, appplatform: PLATFORM.APP });
    const result = await graphql(schema, mutation, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.MeEdit.error).toBe(null);
    expect(result.data?.MeEdit.me.fullName).toBe(name);
    expect(result.data?.MeEdit.me.email).toBe(email);
    expect(sanitizeTestObject(result)).toMatchSnapshot();
  });

  it('should not edit without me', async () => {
    const name = 'Awesome Name';
    const email = 'email@newmail.com';

    const mutation = gql`
      mutation M($input: MeEditInput!) {
        MeEdit(input: $input) {
          me {
            id
            fullName
            email
          }
          error
        }
      }
    `;

    const variables = {
      input: {
        name,
        email,
      },
    };
    const rootValue = {};
    const context = await getContext({ appplatform: PLATFORM.APP });
    const result = await graphql(schema, mutation, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.MeEdit.error).toBe('Unauthorized');
    expect(result.data?.MeEdit.me).toBe(null);
  });

  it('should not edit me with invalid email', async () => {
    const user = await createUser();

    const name = 'jean';
    const email = 'blablabla';

    const mutation = gql`
      mutation M($input: MeEditInput!) {
        MeEdit(input: $input) {
          me {
            id
            email
          }
          error
        }
      }
    `;

    const variables = {
      input: {
        name,
        email,
      },
    };
    const rootValue = {};
    const context = await getContext({ user, appplatform: PLATFORM.APP });
    const result = await graphql(schema, mutation, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.MeEdit.error).toBe('The email must be a valid email.');
    expect(result.data?.MeEdit.me).toBe(null);
  });

  it('should not edit me using password with less than 6 characters', async () => {
    const user = await createUser();

    const name = 'jean';
    const email = 'email@newmail.com';
    const password = '1234';

    const mutation = gql`
      mutation M($input: MeEditInput!) {
        MeEdit(input: $input) {
          me {
            id
            email
          }
          error
        }
      }
    `;

    const variables = {
      input: {
        name,
        email,
        password,
      },
    };
    const rootValue = {};
    const context = await getContext({ user, appplatform: PLATFORM.APP });
    const result = await graphql(schema, mutation, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.MeEdit.error).toBe('Password must be at least 6 characters.');
    expect(result.data?.MeEdit.me).toBe(null);
  });

  it('should not edit me if email belongs to another user', async () => {
    const testUser = await createUser();
    const user = await createUser();

    const name = 'jean';
    const password = '123456';

    const mutation = gql`
      mutation M($input: MeEditInput!) {
        MeEdit(input: $input) {
          me {
            id
            email
          }
          error
        }
      }
    `;

    const variables = {
      input: {
        name,
        email: testUser.email.email,
        password,
      },
    };

    const rootValue = {};
    const context = await getContext({ user, appplatform: PLATFORM.APP });
    const result = await graphql(schema, mutation, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.MeEdit.error).toBe('The email is already associated with another account.');
    expect(result.data?.MeEdit.me).toBe(null);
  });
});
