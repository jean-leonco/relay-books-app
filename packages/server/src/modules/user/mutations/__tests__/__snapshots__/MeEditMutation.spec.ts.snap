// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeEditMutation should edit me 1`] = `
Object {
  "data": Object {
    "MeEdit": Object {
      "error": "EMPTY",
      "me": Object {
        "email": "email@newmail.com",
        "fullName": "Awesome Name",
        "id": "FROZEN-ID",
      },
    },
  },
}
`;
